File: cmd/symdesk/retention.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/danieljustus/symaira-desktop/internal/retention"
	"github.com/danieljustus/symaira-desktop/internal/service"
	"github.com/danieljustus/symaira-desktop/internal/sidecar"
)

func newRetentionCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "retention",
		Short: "Evaluate and apply document retention rules",
	}

	cmd.AddCommand(newRetentionEvalCmd())
	cmd.AddCommand(newRetentionListCmd())
	cmd.AddCommand(newRetentionAcceptCmd())
	cmd.AddCommand(newRetentionRejectCmd())
	cmd.AddCommand(newRetentionDiffCmd())
	cmd.AddCommand(newRetentionHistoryCmd())

	return cmd
}

func newRetentionEvalCmd() *cobra.Command {
	var rulesFile string
	evalCmd := &cobra.Command{
		Use:   "eval",
		Short: "Evaluate retention rules and stage a reviewable proposal",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			vRoot, db, err := initServiceDeps()
			if err != nil {
				return err
			}
			defer closeWithWarning("sidecar database", db.Close)

			if rulesFile == "" {
				rulesFile = filepath.Join(vRoot, ".symdesk", "retention-rules.yaml")
			}

			rules, err := retention.LoadRules(rulesFile)
			if err != nil {
				return fmt.Errorf("load rules from %s: %w", rulesFile, err)
			}

			svc := service.New(vRoot, db)
			docs, err := svc.DocsList(sidecar.DocsFilter{})
			if err != nil {
				return fmt.Errorf("list documents: %w", err)
			}

			now := time.Now()
			var allItems []retention.ProposalItem
			var stateFailures []string

			for _, rule := range rules {
				for _, d := range docs {
					state, stateErr := svc.RetentionState(d.Path)
					if stateErr != nil {
						stateFailures = append(stateFailures, fmt.Sprintf("%s: %v", d.Path, stateErr))
						continue
					}
					// Dataset handles are eligible only for the rule they declare.
					// This check uses the authoritative handle, not sidecar metadata.
					if state.Dataset {
						if rule.Name != state.RuleName {
							continue
						}
					}
					items := retention.Evaluate(rule, []retention.DocMeta{state.Meta}, now)
					for i := range items {
						items[i].RuleName = rule.Name
						items[i].Fingerprint = state.Fingerprint
					}
					allItems = append(allItems, items...)
				}
			}

			if len(stateFailures) > 0 {
				return fmt.Errorf("retention evaluation failed closed: %s", strings.Join(stateFailures, "; "))
			}

			runID := fmt.Sprintf("ret-%d", now.Unix())
			proposal := retention.Proposal{
				RunID:    runID,
				RuleName: "batch",
				Created:  now.UTC(),
				Status:   "pending",
				Items:    allItems,
			}

			if err := retention.WriteProposal(vRoot, proposal); err != nil {
				return err
			}

			return outputResult(map[string]interface{}{
				"status":     "pending",
				"run_id":     runID,
				"item_count": len(allItems),
				"items":      allItems,
			})
		},
	}
	evalCmd.Flags().StringVar(&rulesFile, "rules", "", "path to retention rules YAML file (default: <vault>/.symdesk/retention-rules.yaml)")
	return evalCmd
}

func newRetentionListCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "list",
		Short: "List documents due to expire",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			vRoot, _, err := initServiceDeps()
			if err != nil {
				return err
			}

			dir := retention.ProposalDir(vRoot)
			entries, err := os.ReadDir(dir)
			if err != nil {
				if os.IsNotExist(err) {
					return outputResult(map[string]interface{}{"proposals": []string{}, "message": "no pending proposals"})
				}
				return err
			}

			var proposals []retention.Proposal
			for _, entry := range entries {
				if entry.IsDir() || filepath.Ext(entry.Name()) != ".json" {
					continue
				}
				runID := entry.Name()[:len(entry.Name())-5] // strip .json
				p, err := retention.LoadProposal(vRoot, runID)
				if err != nil {
					continue
				}
				if p.Status != retention.ProposalStatusPending && p.Status != retention.ProposalStatusFailed && p.Status != retention.ProposalStatusPartial {
					continue
				}
				proposals = append(proposals, p)
			}

			if jsonFlag {
				return outputResult(proposals)
			}

			if len(proposals) == 0 {
				fmt.Println("no pending retention proposals")
				return nil
			}

			for _, p := range proposals {
				fmt.Printf("Proposal %s (%s): %d items pending review\n", p.RunID, p.Created.Local().Format("2006-01-02 15:04"), len(p.Items))
				for _, item := range p.Items {
					fmt.Printf("  %s — expires %s → %s\n", item.Path, item.ExpiresAt, item.Action)
				}
			}
			return nil
		},
	}
}

func newRetentionAcceptCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "accept <run-id>",
		Short: "Accept a pending retention proposal",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			vRoot, db, err := initServiceDeps()
			if err != nil {
				return err
			}
			defer closeWithWarning("sidecar database", db.Close)

			p, err := retention.LoadProposal(vRoot, args[0])
			if err != nil {
				return err
			}
			if p.Status != retention.ProposalStatusPending && p.Status != retention.ProposalStatusFailed && p.Status != retention.ProposalStatusPartial && p.Status != retention.ProposalStatusAccepted {
				return fmt.Errorf("proposal %s is %s", args[0], p.Status)
			}

			svc := service.New(vRoot, db)
			now := time.Now()
			acted := 0
			var failures []string

			for i := range p.Items {
				item := &p.Items[i]
				if item.Status == retention.ProposalItemStatusAccepted {
					continue
				}
				if item.Status == retention.ProposalItemStatusActionCompleted {
					if err := retention.AppendHistory(vRoot, retention.HistoryEntry{
						Timestamp: now.UTC(),
						ActionID:  retention.StableActionID(p.RunID, i),
						RuleName:  item.RuleName,
						Action:    item.Action,
						Path:      item.Path,
						Title:     item.Title,
					}); err != nil {
						item.Failure = fmt.Sprintf("history append failed after action: %v", err)
						failures = append(failures, item.Path+": "+item.Failure)
						continue
					}
					item.Status = retention.ProposalItemStatusAccepted
					item.Failure = ""
					p.Status = retentionProposalStatus(p.Items)
					if err := retention.WriteProposal(vRoot, p); err != nil {
						return fmt.Errorf("save retention progress for %s: %w", item.Path, err)
					}
					continue
				}

				state, stateErr := svc.RetentionState(item.Path)
				if stateErr != nil {
					failures = appendRetentionFailure(item, fmt.Sprintf("cannot re-read authoritative state: %v", stateErr), failures)
					continue
				}
				if expectedSlug, isDatasetPath := datasetRetentionSlug(item.Path); isDatasetPath {
					if !state.Dataset || state.Meta.Path != item.Path || state.Meta.Title == "" || expectedSlug == "" {
						failures = appendRetentionFailure(item, "dataset handle is missing or changed", failures)
						continue
					}
					if item.RuleName == "" || state.RuleName != item.RuleName {
						failures = appendRetentionFailure(item, fmt.Sprintf("dataset declares retention rule %q, proposal requires %q", state.RuleName, item.RuleName), failures)
						continue
					}
				}
				if item.Fingerprint == "" {
					failures = appendRetentionFailure(item, "proposal has no fingerprint; re-run retention eval", failures)
					continue
				}
				if state.Fingerprint != item.Fingerprint {
					failures = appendRetentionFailure(item, "proposal is stale: authoritative fingerprint changed", failures)
					continue
				}

				var actionErr error
				if slug, isDatasetPath := datasetRetentionSlug(item.Path); isDatasetPath {
					if item.Action != retention.ActionTrash {
						actionErr = fmt.Errorf("dataset retention action must be trash, got %q", item.Action)
					} else {
						actionErr = svc.DatasetPurge(slug, item.RuleName)
					}
				} else {
					switch item.Action {
					case retention.ActionTrash:
						_, actionErr = svc.NoteDelete(item.Path)
					case retention.ActionFlagReview:
						actionErr = svc.DocStatus(item.Path, "needs_review")
					default:
						actionErr = fmt.Errorf("unsupported retention action %q", item.Action)
					}
				}
				if actionErr != nil {
					failures = appendRetentionFailure(item, fmt.Sprintf("action failed: %v", actionErr), failures)
					continue
				}

				item.Status = retention.ProposalItemStatusActionCompleted
				item.Failure = ""
				p.Status = retentionProposalStatus(p.Items)
				if err := retention.WriteProposal(vRoot, p); err != nil {
					return fmt.Errorf("save action progress for %s: %w", item.Path, err)
				}

				if err := retention.AppendHistory(vRoot, retention.HistoryEntry{
					Timestamp: now.UTC(),
					ActionID:  retention.StableActionID(p.RunID, i),
					RuleName:  item.RuleName,
					Action:    item.Action,
					Path:      item.Path,
					Title:     item.Title,
				}); err != nil {
					item.Failure = fmt.Sprintf("history append failed after action: %v", err)
					failures = append(failures, item.Path+": "+item.Failure)
					continue
				}
				item.Status = retention.ProposalItemStatusAccepted
				item.Failure = ""
				acted++
				p.Status = retentionProposalStatus(p.Items)
				if err := retention.WriteProposal(vRoot, p); err != nil {
					return fmt.Errorf("save retention progress for %s: %w", item.Path, err)
				}
			}

			p.Status = retentionProposalStatus(p.Items)
			if err := retention.WriteProposal(vRoot, p); err != nil {
				return fmt.Errorf("save retention proposal: %w", err)
			}
			result := map[string]interface{}{
				"status":   p.Status,
				"run_id":   args[0],
				"acted":    acted,
				"failures": failures,
				"items":    p.Items,
			}
			if err := outputResult(result); err != nil {
				return err
			}
			if len(failures) > 0 {
				return fmt.Errorf("retention acceptance %s: %s", p.Status, strings.Join(failures, "; "))
			}
			return nil
		},
	}
}

func appendRetentionFailure(item *retention.ProposalItem, message string, failures []string) []string {
	item.Status = ""
	item.Failure = message
	return append(failures, item.Path+": "+message)
}

func retentionProposalStatus(items []retention.ProposalItem) string {
	accepted, failed, pending := 0, 0, 0
	for _, item := range items {
		switch {
		case item.Status == retention.ProposalItemStatusAccepted:
			accepted++
		case item.Failure != "":
			failed++
		default:
			pending++
		}
	}
	if failed == 0 && pending == 0 {
		return retention.ProposalStatusAccepted
	}
	if failed == 0 {
		return retention.ProposalStatusPending
	}
	if accepted == 0 {
		return retention.ProposalStatusFailed
	}
	return retention.ProposalStatusPartial
}

func datasetRetentionSlug(path string) (string, bool) {
	path = filepath.ToSlash(strings.TrimSpace(path))
	if !strings.HasPrefix(path, "datasets/") || !strings.HasSuffix(path, ".md") {
		return "", false
	}
	slug := strings.TrimSuffix(strings.TrimPrefix(path, "datasets/"), ".md")
	if slug == "" || strings.Contains(slug, "/") {
		return "", false
	}
	return slug, true
}

func newRetentionRejectCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "reject <run-id>",
		Short: "Reject a pending retention proposal",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			vRoot, _, err := initServiceDeps()
			if err != nil {
				return err
			}
			p, err := retention.LoadProposal(vRoot, args[0])
			if err != nil {
				return err
			}
			p.Status = "rejected"
			if err := retention.WriteProposal(vRoot, p); err != nil {
				return err
			}
			return outputResult(map[string]string{"status": "rejected", "run_id": args[0]})
		},
	}
}

func newRetentionDiffCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "diff <run-id>",
		Short: "Show the proposed retention actions",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			vRoot, _, err := initServiceDeps()
			if err != nil {
				return err
			}
			p, err := retention.LoadProposal(vRoot, args[0])
			if err != nil {
				return err
			}
			return outputResult(p.Items)
		},
	}
}

func newRetentionHistoryCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "history",
		Short: "Show the history of executed retention actions",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			vRoot, _, err := initServiceDeps()
			if err != nil {
				return err
			}
			entries, err := retention.LoadHistory(vRoot)
			if err != nil {
				return err
			}
			if jsonFlag {
				return outputResult(entries)
			}
			if len(entries) == 0 {
				fmt.Println("no retention actions recorded")
				return nil
			}
			for _, e := range entries {
				fmt.Printf("%s  %s  %s → %s\n",
					e.Timestamp.Local().Format("2006-01-02 15:04:05"),
					e.RuleName,
					e.Path,
					e.Action)
			}
			return nil
		},
	}
}

File: crates/symdesk-cli/src/retention.rs
//! Go: `cmd/symdesk/retention.go`.
//!
//! Retention evaluation and review commands backed by the sidecar and vault
//! state APIs. Acceptance remains out of scope because it requires service
//! mutations.

use std::{
    collections::BTreeMap,
    fs,
    path::{Path, PathBuf},
};

use clap::{Arg, ArgAction, Command};
use serde_json::{Value, json};

use symdesk_index::open_for_vault;
use symdesk_vault::retention::{
    PROPOSAL_STATUS_FAILED, PROPOSAL_STATUS_PARTIAL, PROPOSAL_STATUS_PENDING, Proposal,
    ProposalItem, evaluate, history_path, load_history, load_proposal, load_rules, proposal_dir,
    write_proposal,
};
use symdesk_vault::retention_state::retention_state;
use time::OffsetDateTime;

use crate::{emit_error, write_go_json, write_stdout};

/// Go: `newRetentionCmd`.
pub fn cli() -> Command {
    Command::new("retention")
        .about("Evaluate and apply document retention rules")
        .arg(
            Arg::new("verbose")
                .long("verbose")
                .action(ArgAction::SetTrue),
        )
        .subcommand(
            Command::new("eval")
                .about("Evaluate retention rules and stage a reviewable proposal")
                .arg(
                    Arg::new("rules")
                        .long("rules")
                        .num_args(1)
                        .help("Path to retention rules YAML file (default: <vault>/.symdesk/retention-rules.yaml)"),
                ),
        )
        .subcommand(Command::new("list").about("List documents due to expire"))
        .subcommand(
            Command::new("reject")
                .about("Reject a pending retention proposal")
                .arg(Arg::new("run-id").num_args(0..).action(ArgAction::Append)),
        )
        .subcommand(
            Command::new("diff")
                .about("Show the proposed retention actions")
                .arg(Arg::new("run-id").num_args(0..).action(ArgAction::Append)),
        )
        .subcommand(
            Command::new("history")
                .about("Show the history of executed retention actions")
                .arg(Arg::new("extra").num_args(0..).action(ArgAction::Append)),
        )
}

/// Go: `newRetentionEvalCmd`'s `RunE`.
pub fn run_eval(
    vault: Option<&str>,
    rules_path: Option<&str>,
    output_json: bool,
) -> std::process::ExitCode {
    let vault_root = match crate::resolve_vault(vault) {
        Ok(root) => root,
        Err(error) => return emit_error(error, output_json),
    };
    let sidecar = match open_for_vault(&vault_root) {
        Ok(sidecar) => sidecar,
        Err(error) => return emit_error(error.to_string(), output_json),
    };
    let rules_file = rules_path
        .map(PathBuf::from)
        .unwrap_or_else(|| vault_root.join(".symdesk").join("retention-rules.yaml"));
    let rules = match load_rules(&rules_file) {
        Ok(rules) => rules,
        Err(error) => {
            return emit_error(
                format!("load rules from {}: {error}", rules_file.display()),
                output_json,
            );
        }
    };
    let docs = match sidecar.list_files(&vault_root, "") {
        Ok(docs) => docs,
        Err(error) => return emit_error(format!("list documents: {error}"), output_json),
    };

    let now = OffsetDateTime::now_utc();
    let mut all_items = Vec::new();
    let mut state_failures = Vec::new();
    for rule in &rules {
        for doc in &docs {
            let relative = match Path::new(&doc.path).strip_prefix(&vault_root) {
                Ok(path) => match path.to_str() {
                    Some(path) => path.replace('\\', "/"),
                    None => {
                        state_failures.push(format!("{}: non-UTF-8 vault path", doc.path));
                        continue;
                    }
                },
                Err(_) => {
                    state_failures.push(format!("{}: path is outside vault", doc.path));
                    continue;
                }
            };
            let state = match retention_state(&vault_root, &relative) {
                Ok(state) => state,
                Err(error) => {
                    state_failures.push(format!("{relative}: {error}"));
                    continue;
                }
            };
            if state.dataset && rule.name != state.rule_name {
                continue;
            }
            for mut item in evaluate(rule, &[state.meta], now) {
                item.rule_name.clone_from(&rule.name);
                item.fingerprint.clone_from(&state.fingerprint);
                all_items.push(item);
            }
        }
    }
    if !state_failures.is_empty() {
        return emit_error(
            format!(
                "retention evaluation failed closed: {}",
                state_failures.join("; ")
            ),
            output_json,
        );
    }

    let run_id = format!("ret-{}", now.unix_timestamp());
    let item_count = all_items.len();
    let proposal = Proposal {
        run_id: run_id.clone(),
        rule_name: "batch".to_owned(),
        created: now.to_offset(time::UtcOffset::UTC),
        items: (!all_items.is_empty()).then_some(all_items),
        status: PROPOSAL_STATUS_PENDING.to_owned(),
    };
    if let Err(error) = write_proposal(&vault_root, &proposal) {
        return emit_error(error.to_string(), output_json);
    }
    let result = json!({
        "status": PROPOSAL_STATUS_PENDING,
        "run_id": run_id,
        "item_count": item_count,
        "items": proposal.items,
    });
    output(result, output_json)
}

/// Go: `newRetentionListCmd`'s `RunE`.
pub fn run_list(vault: Option<&str>, output_json: bool) -> std::process::ExitCode {
    let vault_root = match retention_vault(vault, output_json) {
        Ok(root) => root,
        Err(exit) => return exit,
    };

    let dir = proposal_dir(&vault_root);
    let mut entries = match fs::read_dir(&dir) {
        Ok(entries) => entries,
        Err(error) => {
            if error.kind() == std::io::ErrorKind::NotFound {
                // Go: `outputResult(map[string]interface{}{"proposals": []string{}, "message": "no pending proposals"})`
                return output(
                    json!({"proposals": Vec::<String>::new(), "message": "no pending proposals"}),
                    output_json,
                );
            }
            return emit_error(format!("open {}: {error}", dir.display()), output_json);
        }
    };

    // Go's `os.ReadDir` returns entries sorted by file name.
    let mut sorted: Vec<_> = entries.by_ref().flatten().collect();
    sorted.sort_by_key(|entry| entry.file_name());

    let mut proposals: Vec<Proposal> = Vec::new();
    for entry in sorted {
        let name = entry.file_name();
        let name = name.to_string_lossy().to_string();
        let path = entry.path();
        if path.is_dir() || path.extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        let Some(run_id) = name.strip_suffix(".json") else {
            continue;
        };
        let proposal = match load_proposal(&vault_root, run_id) {
            Ok(proposal) => proposal,
            // Go skips an unreadable or undecodable proposal.
            Err(_) => continue,
        };
        if !matches!(
            proposal.status.as_str(),
            PROPOSAL_STATUS_PENDING | PROPOSAL_STATUS_FAILED | PROPOSAL_STATUS_PARTIAL
        ) {
            continue;
        }
        proposals.push(proposal);
    }

    if output_json {
        if proposals.is_empty() {
            return write_stdout("null\n".to_owned());
        }
        return write_go_json(&proposals);
    }

    if proposals.is_empty() {
        return write_stdout("no pending retention proposals\n".to_owned());
    }

    let mut rendered = String::new();
    for proposal in &proposals {
        rendered.push_str(&format!(
            "Proposal {} ({}): {} items pending review\n",
            proposal.run_id,
            go_local_time(proposal.created, false),
            proposal.items.as_ref().map_or(0, Vec::len)
        ));
        for item in proposal.items.as_deref().unwrap_or(&[]) {
            rendered.push_str(&format!(
                "  {} — expires {} → {}\n",
                item.path, item.expires_at, item.action
            ));
        }
    }
    write_stdout(rendered)
}

/// Go: `newRetentionRejectCmd`'s `RunE`.
pub fn run_reject(
    vault: Option<&str>,
    run_ids: &[String],
    output_json: bool,
) -> std::process::ExitCode {
    let run_id = match exact_one(run_ids, output_json) {
        Ok(run_id) => run_id,
        Err(exit) => return exit,
    };
    let vault_root = match retention_vault(vault, output_json) {
        Ok(root) => root,
        Err(exit) => return exit,
    };
    let mut proposal = match load_proposal_for_cli(&vault_root, run_id) {
        Ok(proposal) => proposal,
        Err(error) => return emit_error(error, output_json),
    };
    proposal.status = "rejected".to_owned();
    if let Err(error) = write_proposal(&vault_root, &proposal) {
        return emit_error(error.to_string(), output_json);
    }

    if output_json {
        let result = BTreeMap::from([("run_id", run_id), ("status", "rejected")]);
        write_go_json(&result)
    } else {
        write_stdout(format!("map[run_id:{run_id} status:rejected]\n"))
    }
}

/// Go: `newRetentionDiffCmd`'s `RunE`.
pub fn run_diff(
    vault: Option<&str>,
    run_ids: &[String],
    output_json: bool,
) -> std::process::ExitCode {
    let run_id = match exact_one(run_ids, output_json) {
        Ok(run_id) => run_id,
        Err(exit) => return exit,
    };
    let vault_root = match retention_vault(vault, output_json) {
        Ok(root) => root,
        Err(exit) => return exit,
    };
    let proposal = match load_proposal_for_cli(&vault_root, run_id) {
        Ok(proposal) => proposal,
        Err(error) => return emit_error(error, output_json),
    };

    if output_json {
        return write_go_json(&proposal.items);
    }
    write_stdout(render_items(proposal.items.as_deref().unwrap_or(&[])))
}

/// Go: `newRetentionHistoryCmd`'s `RunE`.
pub fn run_history(
    vault: Option<&str>,
    extra: &[String],
    output_json: bool,
) -> std::process::ExitCode {
    if let Some(extra) = extra.first() {
        return emit_error(
            format!("unknown command {extra:?} for \"symdesk retention history\""),
            output_json,
        );
    }
    let vault_root = match retention_vault(vault, output_json) {
        Ok(root) => root,
        Err(exit) => return exit,
    };
    let path = history_path(&vault_root);
    let missing =
        fs::metadata(&path).is_err_and(|error| error.kind() == std::io::ErrorKind::NotFound);
    let entries = match load_history(&vault_root) {
        Ok(entries) => entries,
        Err(error) => return emit_error(error.to_string(), output_json),
    };

    if output_json {
        if missing {
            return write_stdout("null\n".to_owned());
        }
        return write_go_json(&entries);
    }
    if entries.is_empty() {
        return write_stdout("no retention actions recorded\n".to_owned());
    }

    let mut rendered = String::new();
    for entry in &entries {
        rendered.push_str(&format!(
            "{}  {}  {} → {}\n",
            go_local_time(entry.timestamp, true),
            entry.rule_name,
            entry.path,
            entry.action
        ));
    }
    write_stdout(rendered)
}

fn retention_vault(
    vault: Option<&str>,
    output_json: bool,
) -> Result<PathBuf, std::process::ExitCode> {
    let vault_root = crate::resolve_vault(vault).map_err(|error| emit_error(error, output_json))?;
    // Go discards these `*sidecar.DB` handles without closing them for list,
    // reject, diff, and history. Keep the Rust connection alive until process
    // exit so metadata.json plus sidecar.db-wal/sidecar.db-shm match Go.
    let sidecar =
        open_for_vault(&vault_root).map_err(|error| emit_error(error.to_string(), output_json))?;
    std::mem::forget(sidecar);
    Ok(vault_root)
}

fn exact_one(values: &[String], output_json: bool) -> Result<&str, std::process::ExitCode> {
    if values.len() != 1 {
        return Err(emit_error(
            format!("accepts 1 arg(s), received {}", values.len()),
            output_json,
        ));
    }
    Ok(values[0].as_str())
}

fn load_proposal_for_cli(vault_root: &Path, run_id: &str) -> Result<Proposal, String> {
    load_proposal(vault_root, run_id).map_err(|error| error.to_string())
}

fn render_items(items: &[ProposalItem]) -> String {
    let rendered = items
        .iter()
        .map(|item| {
            format!(
                "{{Path:{} Title:{} ReferenceDate:{} ExpiresAt:{} Action:{} RuleName:{} Fingerprint:{} Status:{} Failure:{}}}",
                item.path,
                item.title,
                item.reference_date,
                item.expires_at,
                item.action,
                item.rule_name,
                item.fingerprint,
                item.status,
                item.failure
            )
        })
        .collect::<Vec<_>>()
        .join(" ");
    format!("[{rendered}]\n")
}

/// Go: `.Local().Format("2006-01-02 15:04[:05]")`; resolve the offset at the
/// event instant so historical daylight-saving transitions match.
fn go_local_time(value: time::OffsetDateTime, seconds: bool) -> String {
    let offset = time::UtcOffset::local_offset_at(value).unwrap_or(time::UtcOffset::UTC);
    let local = value.to_offset(offset);
    if seconds {
        format!(
            "{:04}-{:02}-{:02} {:02}:{:02}:{:02}",
            local.year(),
            u8::from(local.month()),
            local.day(),
            local.hour(),
            local.minute(),
            local.second()
        )
    } else {
        format!(
            "{:04}-{:02}-{:02} {:02}:{:02}",
            local.year(),
            u8::from(local.month()),
            local.day(),
            local.hour(),
            local.minute()
        )
    }
}

fn output(value: Value, output_json: bool) -> std::process::ExitCode {
    if output_json {
        write_go_json(&value)
    } else {
        // Go prints the map with `fmt.Printf("%+v\n", data)`.
        let rendered = match &value {
            Value::Object(map) => {
                let inner = map
                    .iter()
                    .map(|(key, value)| format!("{key}:{}", go_value(value)))
                    .collect::<Vec<_>>()
                    .join(" ");
                format!("map[{inner}]")
            }
            other => go_value(other),
        };
        write_stdout(format!("{rendered}\n"))
    }
}

fn go_value(value: &Value) -> String {
    match value {
        Value::String(text) => text.clone(),
        Value::Array(items) if items.is_empty() => "[]".to_owned(),
        Value::Array(items) => format!(
            "[{}]",
            items.iter().map(go_value).collect::<Vec<_>>().join(" ")
        ),
        Value::Null => "<nil>".to_owned(),
        other => other.to_string(),
    }
}
